cmd/bank/main.go

```diff
@@ -1,23 +1,121 @@
 package main
 
 import (
+	"bufio"
 	"context"
+	"database/sql"
+	"fmt"
+	"math/rand"
+	"os"
+	"strconv"
+	"strings"
 	"time"
 
-	"github.com/dogmatiq/dogma"
 	"github.com/dogmatiq/example"
 	"github.com/dogmatiq/example/database"
 	"github.com/dogmatiq/example/messages"
 	"github.com/dogmatiq/example/messages/commands"
 	"github.com/dogmatiq/testkit/engine"
+	"github.com/gdamore/tcell"
+	"github.com/google/uuid"
 	_ "github.com/mattn/go-sqlite3"
+	"github.com/rivo/tview"
 )
 
+const refreshInterval = 1000 * time.Millisecond
+
+func currentTimeString(t time.Time) string {
+	return fmt.Sprintf(t.Format("Date: 2006-01-02 15:04:05"))
+}
+
+func updateTime(as *appState) {
+	for {
+		time.Sleep(refreshInterval)
+		as.guiApp.QueueUpdateDraw(func() {
+			as.time = as.time.Add(time.Minute) // accelerated example bank time
+			as.clockView.SetText(currentTimeString(as.time))
+
+			as.engine.Tick(
+				context.Background(),
+				engine.WithCurrentTime(as.time),
+			)
+		})
+	}
+}
+
 func businessDayFromTime(t time.Time) string {
 	return t.Format(messages.BusinessDateFormat)
 }
 
+func readString(reader *bufio.Reader) string {
+	text, _ := reader.ReadString('\n')
+	return strings.TrimSuffix(text, "\n")
+}
+
+func generateID() string {
+	return uuid.New().String()
+}
+
+func generateCustomerNumber() string {
+	return fmt.Sprintf("%d%03d", rand.Intn(8)+1, rand.Intn(900))
+}
+
+func generateAccountNumber() string {
+	return fmt.Sprintf("%d%04d", rand.Intn(8)+1, rand.Intn(9000))
+}
+
+type appState struct {
+	db     *sql.DB
+	engine *engine.Engine
+	reader *bufio.Reader
+	time   time.Time
+
+	// ui ...
```

Comment:
I doubt that I need to store all these now. Most that are recreated each time could just stay in the scope they are defined in?

Reply:
Removed.